File: apps/lib/services/external/ohana_client.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../../models/program.dart';

/// Client for the Ohana API (SMC-Connect)
/// HSDS-compliant API for San Mateo County services
/// https://api.smc-connect.org/
class OhanaClient {
  static const String baseUrl = 'https://api.smc-connect.org';

  final http.Client _client;

  OhanaClient({http.Client? client}) : _client = client ?? http.Client();

  /// Search for services by keyword
  Future<List<Program>> search(String keyword) async {
    final uri = Uri.parse('$baseUrl/search').replace(
      queryParameters: {
        'keyword': keyword,
        'per_page': '100',
      },
    );

    try {
      final response = await _client.get(uri, headers: _headers);

      if (response.statusCode != 200) {
        return [];
      }

      final List<dynamic> data = json.decode(response.body);
      return data.map((item) => _mapToProgram(item)).toList();
    } catch (e) {
      return [];
    }
  }

  /// Fetch organizations (paginated)
  Future<List<Program>> fetchOrganizations({int page = 1, int perPage = 50}) async {
    final uri = Uri.parse('$baseUrl/organizations').replace(
      queryParameters: {
        'page': page.toString(),
        'per_page': perPage.toString(),
      },
    );

    try {
      final response = await _client.get(uri, headers: _headers);

      if (response.statusCode != 200) {
        return [];
      }

      final List<dynamic> data = json.decode(response.body);
      return data.map((item) => _mapOrganizationToProgram(item)).toList();
    } catch (e) {
      return [];
    }
  }

  /// Fetch all available programs from Ohana API
  /// Uses multiple searches to cover common categories
  Future<List<Program>> fetchAll() async {
    final searchTerms = [
      'food',
      'housing',
      'shelter',
      'health',
      'employment',
      'legal',
      'education',
    ];

    final results = await Future.wait(
      searchTerms.map((term) => search(term)),
    );

    // Combine and deduplicate by external ID
    final uniquePrograms = <String, Program>{};
    for (final list in results) {
      for (final program in list) {
        final key = program.externalId ?? program.id;
        uniquePrograms[key] = program;
      }
    }

    return uniquePrograms.values.toList();
  }

  Map<String, String> get _headers => {
        'Accept': 'application/json',
        'User-Agent': 'BayNavigator/1.0',
      };

  /// Map Ohana search result to Program model
  Program _mapToProgram(Map<String, dynamic> item) {
    final organization = item['organization'] as Map<String, dynamic>?;
    final location = item['location'] as Map<String, dynamic>?;
    final services = item['services'] as List<dynamic>?;

    final orgName = organization?['name'] as String? ?? 'Unknown Organization';
    final serviceName = services?.isNotEmpty == true
        ? services!.first['name'] as String?
        : null;

    final name = serviceName != null ? '$orgName - $serviceName' : orgName;

    // Extract address
    final address = location?['address'] as Map<String, dynamic>?;
    final fullAddress = _buildAddress(address);

    // Extract phone
    final phones = location?['phones'] as List<dynamic>?;
    final phone = phones?.isNotEmpty == true
        ? phones!.first['number'] as String?
        : null;

    // Extract coordinates
    final lat = _parseDouble(location?['latitude']);
    final lng = _parseDouble(location?['longitude']);

    // Extract description from first service
    final description = services?.isNotEmpty == true
        ? services!.first['description'] as String? ?? 'Service provided by $orgName.'
        : 'Organization providing community services.';

    // Extract eligibility
    final eligibility = services?.isNotEmpty == true
        ? services!.first['eligibility'] as String?
        : null;
    final groups = _parseEligibility(eligibility);

    // Determine category
    final category = _determineCategory(item);

    return Program(
      id: 'ohana-${item['id'] ?? DateTime.now().millisecondsSinceEpoch}',
      name: name,
      category: category,
      description: description,
      groups: groups,
      areas: ['san-mateo'],
      city: address?['city'] as String?,
      website: organization?['website'] as String? ?? 'https://www.smc-connect.org',
      phone: phone,
      address: fullAddress,
      lastUpdated: DateTime.now().toIso8601String().split('T').first,
      latitude: lat,
      longitude: lng,
      dataSource: DataSource.ohana,
      externalId: item['id']?.toString(),
      sourceUrl: 'https://www.smc-connect.org/locations/${item['id']}',
    );
  }

  /// Map organization to Program model
  Program _mapOrganizationToProgram(Map<String, dynamic> item) {
    final name = item['name'] as String? ?? 'Unknown Organization';
    final description = item['description'] as String? ??
        'Organization providing community services in San Mateo County.';

    final locations = item['locations'] as List<dynamic>?;
    final firstLocation = locations?.isNotEmpty == true
        ? locations!.first as Map<String, dynamic>
        : null;

    final address = firstLocation?['address'] as Map<String, dynamic>?;
    final fullAddress = _buildAddress(address);

    final phones = firstLocation?['phones'] as List<dynamic>?;
    final phone = phones?.isNotEmpty == true
        ? phones!.first['number'] as String?
        : null;

    return Program(
      id: 'ohana-org-${item['id'] ?? DateTime.now().millisecondsSinceEpoch}',
      name: name,
      category: 'other',
      description: description,
      groups: [],
      areas: ['san-mateo'],
      city: address?['city'] as String?,
      website: item['website'] as String? ?? 'https://www.smc-connect.org',
      phone: phone,
      address: fullAddress,
      lastUpdated: DateTime.now().toIso8601String().split('T').first,
      dataSource: DataSource.ohana,
      externalId: item['id']?.toString(),
      sourceUrl: 'https://www.smc-connect.org/organizations/${item['id']}',
    );
  }

  String? _buildAddress(Map<String, dynamic>? address) {
    if (address == null) return null;

    final parts = <String>[];
    final street = address['address_1'] as String?;
    final city = address['city'] as String?;
    final state = address['state_province'] as String?;
    final zip = address['postal_code'] as String?;

    if (street != null && street.isNotEmpty) parts.add(street);
    if (city != null && city.isNotEmpty) {
      final cityLine = [city, state, zip].where((p) => p != null && p.isNotEmpty).join(' ');
      parts.add(cityLine);
    }

    return parts.isNotEmpty ? parts.join(', ') : null;
  }

  List<String> _parseEligibility(String? eligibility) {
    if (eligibility == null || eligibility.isEmpty) return [];

    final groups = <String>[];
    final lower = eligibility.toLowerCase();

    if (lower.contains('senior') || lower.contains('older adult') || lower.contains('65+')) {
      groups.add('seniors');
    }
    if (lower.contains('veteran')) {
      groups.add('veterans');
    }
    if (lower.contains('youth') || lower.contains('child') || lower.contains('teen')) {
      groups.add('youth');
    }
    if (lower.contains('disabled') || lower.contains('disability')) {
      groups.add('disabled');
    }
    if (lower.contains('immigrant') || lower.contains('refugee')) {
      groups.add('immigrants');
    }
    if (lower.contains('family') || lower.contains('families')) {
      groups.add('families');
    }
    if (lower.contains('homeless') || lower.contains('unhoused')) {
      groups.add('unhoused');
    }

    return groups;
  }

  String _determineCategory(Map<String, dynamic> item) {
    final services = item['services'] as List<dynamic>?;
    if (services == null || services.isEmpty) return 'other';

    final firstService = services.first as Map<String, dynamic>;
    final name = (firstService['name'] as String? ?? '').toLowerCase();
    final description = (firstService['description'] as String? ?? '').toLowerCase();
    final combined = '$name $description';

    if (combined.contains('food') || combined.contains('meal') || combined.contains('pantry')) {
      return 'food';
    }
    if (combined.contains('housing') || combined.contains('shelter') || combined.contains('rent')) {
      return 'housing';
    }
    if (combined.contains('health') || combined.contains('medical') || combined.contains('clinic')) {
      return 'healthcare';
    }
    if (combined.contains('job') || combined.contains('employment') || combined.contains('work')) {
      return 'employment';
    }
    if (combined.contains('legal') || combined.contains('attorney') || combined.contains('law')) {
      return 'legal';
    }
    if (combined.contains('education') || combined.contains('school') || combined.contains('training')) {
      return 'education';
    }
    if (combined.contains('money') || combined.contains('financial') || combined.contains('cash')) {
      return 'financial';
    }

    return 'other';
  }

  double? _parseDouble(dynamic value) {
    if (value == null) return null;
    if (value is double) return value;
    if (value is int) return value.toDouble();
    if (value is String) return double.tryParse(value);
    return null;
  }

  void dispose() {
    _client.close();
  }
}

File: apps/apple/BayNavigatorCore/Sources/BayNavigatorCore/Services/External/OhanaClient.swift
import Foundation

/// Client for the Ohana API (SMC-Connect)
/// HSDS-compliant API for San Mateo County services
/// https://api.smc-connect.org/
public actor OhanaClient {
    public static let shared = OhanaClient()

    private static let baseUrl = "https://api.smc-connect.org"

    private let urlSession: URLSession

    public init(urlSession: URLSession = .shared) {
        self.urlSession = urlSession
    }

    /// Search for services by keyword
    public func search(keyword: String) async -> [Program] {
        var components = URLComponents(string: "\(Self.baseUrl)/search")
        components?.queryItems = [
            URLQueryItem(name: "keyword", value: keyword),
            URLQueryItem(name: "per_page", value: "100"),
        ]

        guard let url = components?.url else { return [] }

        do {
            var request = URLRequest(url: url)
            request.setValue("application/json", forHTTPHeaderField: "Accept")
            request.setValue("BayNavigator/1.0", forHTTPHeaderField: "User-Agent")

            let (data, response) = try await urlSession.data(for: request)

            guard let httpResponse = response as? HTTPURLResponse,
                  httpResponse.statusCode == 200 else {
                return []
            }

            let items = try JSONDecoder().decode([OhanaSearchResult].self, from: data)
            return items.map { mapToProgram($0) }
        } catch {
            return []
        }
    }

    /// Fetch all available programs from Ohana API
    /// Uses multiple searches to cover common categories
    public func fetchAll() async -> [Program] {
        let searchTerms = ["food", "housing", "shelter", "health", "employment", "legal", "education"]

        var uniquePrograms: [String: Program] = [:]

        for term in searchTerms {
            let results = await search(keyword: term)
            for program in results {
                let key = program.externalId ?? program.id
                uniquePrograms[key] = program
            }
        }

        return Array(uniquePrograms.values)
    }

    private func mapToProgram(_ item: OhanaSearchResult) -> Program {
        let orgName = item.organization?.name ?? "Unknown Organization"
        let serviceName = item.services?.first?.name

        let name = serviceName != nil ? "\(orgName) - \(serviceName!)" : orgName

        let address = buildAddress(item.location?.address)
        let phone = item.location?.phones?.first?.number

        let lat = item.location?.latitude
        let lng = item.location?.longitude

        let description = item.services?.first?.description ?? "Service provided by \(orgName)."

        let eligibility = item.services?.first?.eligibility
        let groups = parseEligibility(eligibility)

        let category = determineCategory(item)

        return Program(
            id: "ohana-\(item.id ?? UUID().hashValue)",
            name: name,
            category: category,
            description: description,
            groups: groups,
            areas: ["san-mateo"],
            city: item.location?.address?.city,
            website: item.organization?.website ?? "https://www.smc-connect.org",
            phone: phone,
            address: address,
            lastUpdated: dateString(),
            latitude: lat,
            longitude: lng,
            dataSource: .ohana,
            externalId: item.id.map(String.init),
            sourceUrl: item.id.map { "https://www.smc-connect.org/locations/\($0)" }
        )
    }

    private func buildAddress(_ address: OhanaAddress?) -> String? {
        guard let address = address else { return nil }

        var parts: [String] = []
        if let street = address.address1, !street.isEmpty {
            parts.append(street)
        }
        if let city = address.city, !city.isEmpty {
            var cityLine = city
            if let state = address.stateProvince, !state.isEmpty {
                cityLine += " \(state)"
            }
            if let zip = address.postalCode, !zip.isEmpty {
                cityLine += " \(zip)"
            }
            parts.append(cityLine)
        }

        return parts.isEmpty ? nil : parts.joined(separator: ", ")
    }

    private func parseEligibility(_ eligibility: String?) -> [String] {
        guard let eligibility = eligibility?.lowercased() else { return [] }

        var groups: [String] = []

        if eligibility.contains("senior") || eligibility.contains("older adult") || eligibility.contains("65+") {
            groups.append("seniors")
        }
        if eligibility.contains("veteran") {
            groups.append("veterans")
        }
        if eligibility.contains("youth") || eligibility.contains("child") || eligibility.contains("teen") {
            groups.append("youth")
        }
        if eligibility.contains("disabled") || eligibility.contains("disability") {
            groups.append("disabled")
        }
        if eligibility.contains("immigrant") || eligibility.contains("refugee") {
            groups.append("immigrants")
        }
        if eligibility.contains("family") || eligibility.contains("families") {
            groups.append("families")
        }
        if eligibility.contains("homeless") || eligibility.contains("unhoused") {
            groups.append("unhoused")
        }

        return groups
    }

    private func determineCategory(_ item: OhanaSearchResult) -> String {
        guard let service = item.services?.first else { return "other" }

        let name = (service.name ?? "").lowercased()
        let description = (service.description ?? "").lowercased()
        let combined = "\(name) \(description)"

        if combined.contains("food") || combined.contains("meal") || combined.contains("pantry") {
            return "food"
        }
        if combined.contains("housing") || combined.contains("shelter") || combined.contains("rent") {
            return "housing"
        }
        if combined.contains("health") || combined.contains("medical") || combined.contains("clinic") {
            return "healthcare"
        }
        if combined.contains("job") || combined.contains("employment") || combined.contains("work") {
            return "employment"
        }
        if combined.contains("legal") || combined.contains("attorney") || combined.contains("law") {
            return "legal"
        }
        if combined.contains("education") || combined.contains("school") || combined.contains("training") {
            return "education"
        }
        if combined.contains("money") || combined.contains("financial") || combined.contains("cash") {
            return "financial"
        }

        return "other"
    }

    private func dateString() -> String {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        return formatter.string(from: Date())
    }
}

// MARK: - Ohana API Response Models

private struct OhanaSearchResult: Codable {
    let id: Int?
    let organization: OhanaOrganization?
    let location: OhanaLocation?
    let services: [OhanaService]?
}

private struct OhanaOrganization: Codable {
    let name: String?
    let website: String?
}

private struct OhanaLocation: Codable {
    let address: OhanaAddress?
    let phones: [OhanaPhone]?
    let latitude: Double?
    let longitude: Double?
}

private struct OhanaAddress: Codable {
    let address1: String?
    let city: String?
    let stateProvince: String?
    let postalCode: String?

    enum CodingKeys: String, CodingKey {
        case address1 = "address_1"
        case city
        case stateProvince = "state_province"
        case postalCode = "postal_code"
    }
}

private struct OhanaPhone: Codable {
    let number: String?
}

private struct OhanaService: Codable {
    let name: String?
    let description: String?
    let eligibility: String?
}
